### DISCUSSION TITLE
Pulling many docker images is significantly slow (docker-from-docker)

### DISCUSSION THREAD

**Discussion by chesleybrown** [community/community#33619] [Codespaces]
I'm trying to orchestrate a large micro service based application (100+ services) within a single Codespace by using [docker-from-docker](https://github.com/danielfdickinson/microsoft-vscode-dev-containers/tree/main/containers/docker-from-docker) and docker compose.

When I run this stack using an AWS EC2 instance, it can be brought up in about 20 minutes. The instance has 4 CPUs and 16GB of memory. It is fairly quick to pull everything down, configure all the services and start them up. The EC2 instance remains performant with all the services running. Worth noting that we do _not_ have to use `docker-from-docker` and are running the docker commands directly on the host.

When I try to do the same process on a Codespace, it is substantially slower. It's taking about 1.5 hours to pull down all the images. And worse again, when I spin up all the services, it consumes most of the CPU and brings the Codespace to a halt. This is using a 8 CPU 16GB instance. I get better results when using the 16 CPU 32 GB instance in that it can run the full stack without using up all the resources, however it is still noticeably slower than the EC2 instance.

I'm looking for any guidance on _why_ this setup would be so much slower and less performant on Codespaces. Even just understanding why pulling the images down can be so much slower (~15 minutes on EC2 instance vs ~1.5 hours on Codespaces) would be a huge help.

**Comment by bjbennet** [comment]
Hi @chesleybrown, there could be a number of reasons why you are experiencing faster performance on AWS compared to Codespaces. A few points that come to mind:
1. Are your images stored in AWS? If so, then that would give you faster performance when running your application on AWS.
2. What region are your images in and what region are you creating Codespaces in? The greater the distance between the region where your images are stored and the region where you are creating Codespaces can be a performance factor.
3. Lastly, on the Codespaces machine during the image pull you could be experiencing a bottleneck with regard to IOPS and bandwidth as we have seen this in the past.

Hopefully, this helps shed some light, on performance factors.

**Reply by chesleybrown** [reply]
Thanks for taking the time to reply!

1. The images are stored on quay.io. I think it's worth noting that I can also download the images to my local MacBook significantly faster than Codespaces. I don't think it's a location issue.
2. Everything was tested using instances in us-east-1.
3. Anyway to determine if this is the issue? What are the IOPS for the Codespace instances?

Thanks again!
